src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTiming.java

```diff
@@ -301,9 +302,25 @@ public static GenericStatus computeChunkStatus2(@Nonnull WorkflowRun run,
                 return GenericStatus.FAILURE;
             }
         }
+        WarningAction warning = findWarningBetween(firstNode, lastNode);
+        if (warning != null) {
+            return GenericStatus.UNSTABLE;
+        }
 
         // Previous chunk before end. If flow continued beyond this, it didn't fail.
-        return (run.getResult() == Result.UNSTABLE) ? GenericStatus.UNSTABLE : GenericStatus.SUCCESS;
+        return GenericStatus.SUCCESS;
+    }
+
+    private static @CheckForNull WarningAction findWarningBetween(@Nonnull FlowNode start, @Nonnull FlowNode end) {
+        // TODO: Cache the result?
+        FlowNode found = new DepthFirstScanner().findFirstMatch(
+                Collections.singletonList(end),
+                Collections.singletonList(start),
+                tempNode -> tempNode.getPersistentAction(WarningAction.class) != null);
+        if (found != null) {
+            return found.getPersistentAction(WarningAction.class);
```

Comment:
Awkward to have to look up the action again. Does `FlowNodeScanner` not return a `Stream` or something that can be easily converted to it? Then you can use `map`, `Objects::nonNull`, and `findFirst`.

Reply:
The helper methods all return a `List`, but `FlowNodeScanner` itself is an `Iterable` so we can turn that into a stream via `.spliterator()`. Done in 5bc959a.